### PR TITLE
Added a github workflow for running examples as integration tests

### PR DESCRIPTION
**Pull request by x3c41a**

### CODE REVIEW

**Comment by x3c41a** [comment]
Waiting for https://github.com/paritytech/polkadot-bulletin-chain/actions/runs/20095449181 to finish...

File: .github/workflows/smoldot-test.yml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+      - name: Install just
+        run: cargo install just --locked || true
+
+      - name: Download zombienet
+        run: |
+          wget -q "https://github.com/paritytech/zombienet/releases/download/v1.3.133/zombienet-linux-x64"
+          chmod +x zombienet-linux-x64
+
+      - name: Run smoldot workflow test
```
Comment (bkontur) [inline .github/workflows/smoldot-test.yml]: We need to run all the examples

smoldot-test.yml -> integration-tests.yml

File: .github/workflows/integration-test.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
+name: Integration Tests
+
+on:
+  push:
+    branches: [main, ci_cd]
+  pull_request:
```
Comment (x3c41a) [inline .github/workflows/integration-test.yml]: I am not a GH CI/CD export, I was just looking for a way to run the workflow from my branch. Now that it's working, I can simply remove my (`ci_cd`) branch from here and merge, right? @bkontur

File: .github/workflows/integration-test.yml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
+
+env:
+  CI_IMAGE: "paritytech/ci-unified:bullseye-1.88.0-2025-06-27-v202507112050"
+
+jobs:
+  smoldot-docker-test:
+    name: Smoldot Docker Test
```
Comment (bkontur) [inline .github/workflows/integration-test.yml]:
```suggestion
  integration-tests:
    name: Integration Tests
```

File: .github/workflows/integration-test.yml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      - name: Install just
+        run: cargo install just --locked || true
+
+      - name: Download zombienet
+        run: |
+          wget -q "https://github.com/paritytech/zombienet/releases/download/v1.3.133/zombienet-linux-x64"
+          chmod +x zombienet-linux-x64
+          echo "ZOMBIENET_BINARY=$GITHUB_WORKSPACE/zombienet-linux-x64" >> $GITHUB_ENV
+
+      - name: Run smoldot workflow test
```
Comment (bkontur) [inline .github/workflows/integration-test.yml]:
```suggestion
      - name: Run authorize and store (PAPI, smoldot)
```

File: .github/workflows/integration-test.yml [hunk 5]
```diff
@@ -0,0 +1,62 @@
+name: Integration Tests
+
+on:
+  push:
+    branches: [main, ci_cd]
```
Comment (bkontur) [inline .github/workflows/integration-test.yml]:
```suggestion
    branches: ["main", "release-*"]
```

File: examples/justfile [hunk 6]
```diff
@@ -261,8 +282,17 @@ ipfs-docker-start: _check-docker
     # Pull latest kubo image if not present
     docker pull ipfs/kubo:latest
     
+    # Determine network mode based on OS
+    if [[ "$OSTYPE" == "darwin"* ]]; then
+        # macOS - use port mapping
+        NETWORK_ARGS="-p 4001:4001 -p 8080:8080 -p 5001:5001"
+    else
+        # Linux (including CI) - use host networking for direct access
+        NETWORK_ARGS="--network host"
```
Comment (bkontur) [inline examples/justfile]: Why `--network host`? Aren't we missing ports for linux?

```suggestion
        NETWORK_ARGS="--network host -p 4001:4001 -p 8080:8080 -p 5001:5001"
```

File: .github/workflows/integration-test.yml [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+      - name: Rust cache
+        uses: Swatinem/rust-cache@v2
+        with:
+          cache-on-failure: true
+          cache-all-crates: true
+
+      - name: Setup Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: '20'
```
Comment (bkontur) [inline .github/workflows/integration-test.yml]:
```suggestion
          node-version: '22'
```

File: .github/workflows/integration-test.yml [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+      - name: Download zombienet
+        run: |
+          wget "https://github.com/paritytech/zombienet/releases/download/v1.3.133/zombienet-linux-x64" || exit 1
+          chmod +x zombienet-linux-x64
+          echo "ZOMBIENET_BINARY=$GITHUB_WORKSPACE/zombienet-linux-x64" >> $GITHUB_ENV
+
+      - name: Run authorize and store (PAPI, smoldot)
+        working-directory: examples
+        run: just run-authorize-and-store-smoldot-docker
```
Comment (bkontur) [inline .github/workflows/integration-test.yml]: I think `docker` or local kubo could be an argument

```suggestion
        run: just run-authorize-and-store-smoldot
```

Comment (bkontur) [reply .github/workflows/integration-test.yml]: > If we keep local set-up, it will make us spend some extra time on maintenance and resolve different corner cases bugs in the future

for CI I am definitely fine with Docker, I am also using docker for local stuff, we don't need to support local Kubo, let's simply remove local Kubo support

File: .github/workflows/integration-test.yml [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+          cache-dependency-path: examples/package.json
+
+      - name: Install just
+        run: cargo install just --locked || true
+
+      - name: Download zombienet
+        run: |
+          wget "https://github.com/paritytech/zombienet/releases/download/v1.3.133/zombienet-linux-x64" || exit 1
+          chmod +x zombienet-linux-x64
+          echo "ZOMBIENET_BINARY=$GITHUB_WORKSPACE/zombienet-linux-x64" >> $GITHUB_ENV
```
Comment (bkontur) [inline .github/workflows/integration-test.yml]: @x3c41a can you please try this? According to the AI the `${{ github.token }}` should eliminate rate limiting issue.
Otherwise, we can go with some caching and/or custom docker image or docker zombienet version later.

```suggestion
      - name: Download zombienet
        run: |
          curl -L \
            -H "Authorization: token ${{ github.token }}" \
            -o zombienet-linux-x64 \
            "https://github.com/paritytech/zombienet/releases/download/v1.3.138/zombienet-linux-x64"
          chmod +x zombienet-linux-x64
          echo "ZOMBIENET_BINARY=$GITHUB_WORKSPACE/zombienet-linux-x64" >> $GITHUB_ENV
```

File: .github/workflows/integration-test.yml [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+      - name: Run authorize and store (PAPI, smoldot)
+        working-directory: examples
+        run: just run-authorize-and-store-smoldot-docker
+
+      - name: Run authorize and store (PAPI, RPC node)
+        working-directory: examples
+        run: just run-authorize-and-store-docker
```
Comment (bkontur) [inline .github/workflows/integration-test.yml]: @x3c41a the last change, let's go with this "docker" suffix, anyway it is just for Kubo, and run Kubo as docker as we discussed

```suggestion
        run: just run-authorize-and-store-smoldot

      - name: Run authorize and store (PAPI, RPC node)
        working-directory: examples
        run: just run-authorize-and-store
```

File: examples/justfile [hunk 11]
```diff
@@ -88,139 +36,34 @@ bulletin-solo-zombienet-start:
...
+    POLKADOT_BULLETIN_BINARY_PATH=$BULLETIN_BINARY $ZOMBIENET_BIN -p native spawn $ZOMBIENET_CONFIG > /tmp/zombienet.log 2>&1 &
+    ZOMBIENET_PID=$!
+    echo $ZOMBIENET_PID > {{ PID_DIR }}/zombienet.pid
+    echo "   Zombienet PID: $ZOMBIENET_PID"
+    echo "   Log: /tmp/zombienet.log"
+    echo "   Waiting for chain to start (15 seconds)..."
+    sleep 15
 
 # Run smoldot example with local Kubo
 run-authorize-and-store-smoldot: build npm-install
```
[the remark this answers is not in the document]
Comment (x3c41a) [reply examples/justfile]: yeah, I parametrised `run-authorize-and-store` and the duplicate should be removed as well
